ports/zephyr/machine_spi.c

```diff
@@ -0,0 +1,217 @@
+#define SPI_LOOP                (0)    //For testing, enable loop mode by setting SPI_LOOP (1)
+
+typedef struct _machine_hard_spi_obj_t {
+    mp_obj_base_t base;
+    const struct device *dev;
+    struct spi_config config;
+} machine_hard_spi_obj_t;
+
+STATIC void machine_hard_spi_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_hard_spi_obj_t *self = self_in;
+    mp_printf(print, "Device: %s | Baudrate: %u | Polarity: %u | Phase: %u | Bits: %u | Firstbit: %s",
+            self->dev->name,
+            self->config.frequency,
+            (self->config.operation & 0x2) >> 1,
+            (self->config.operation & 0x4) >> 2,
+            (self->config.operation & ~0x1F) >> 5,
+            ((self->config.operation & 0x10) >> 4) == 0 ? "MSB" : "LSB");
+}
+
+mp_obj_t machine_hard_spi_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+
+//  MP_MACHINE_SPI_CHECK_FOR_LEGACY_SOFTSPI_CONSTRUCTION(n_args, n_kw, all_args);
+
+    enum {ARG_id, ARG_baudrate, ARG_polarity, ARG_phase, ARG_bits, ARG_firstbit, ARG_sck, ARG_mosi, ARG_miso};
+
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_baudrate, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_BAUDRATE} },
+        { MP_QSTR_polarity, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_POLARITY} },
+        { MP_QSTR_phase, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_PHASE} },
+        { MP_QSTR_bits, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_BITS} },
+        { MP_QSTR_firstbit, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_SPI_FIRSTBIT} }, 
+        { MP_QSTR_sck, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_mosi, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_miso, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    const char *dev_name = mp_obj_str_get_str(args[ARG_id].u_obj);
+    const struct device *dev = device_get_binding(dev_name);
+
+    if (dev == NULL) {
+        mp_raise_ValueError(MP_ERROR_TEXT("device not found"));
+    }
+
+    if ((args[ARG_sck].u_obj != MP_OBJ_NULL) || (args[ARG_miso].u_obj != MP_OBJ_NULL) || (args[ARG_mosi].u_obj != MP_OBJ_NULL)) {
+        mp_raise_NotImplementedError(MP_ERROR_TEXT("explicit choice of sck/miso/mosi is not implemented"));
+    }
+
+    struct spi_config cfg = {
+        .frequency = args[ARG_baudrate].u_int,
+        .operation = SPI_OP_MODE_MASTER |
+                args[ARG_polarity].u_int << 1 |
+                args[ARG_phase].u_int << 2 |
+                SPI_LOOP << 3 |
+                args[ARG_firstbit].u_int << 4 |
+                args[ARG_bits].u_int << 5 |
+                SPI_LINES_SINGLE,
+        .slave = 0,
+        .cs = NULL
```

Comment:
The code formatting check requires that lines 102-109 be indented to the same level (lines 65-72 in the formatting build), but this would put the `.slave` and `.cs` lines at a different indent level than `.frequency` and `.operation`. 

@dpgeorge what do you recommend for formatting here?

Reply:
Try putting parenthesis around the `.operation` argument: `.operation = (...)`.

Reply:
This is the best I could do:
```
    struct spi_config cfg = {
        .frequency = args[ARG_baudrate].u_int,
        .operation = (SPI_OP_MODE_MASTER |
            args[ARG_polarity].u_int << 1 |
                args[ARG_phase].u_int << 2 |
                SPI_LOOP << 3 |
                args[ARG_firstbit].u_int << 4 |
                args[ARG_bits].u_int << 5 |
                SPI_LINES_SINGLE),
        .slave = 0,
        .cs = NULL
    };
```